src/ast/macros.rs

```diff
@@ -0,0 +1,327 @@
+            range: Option<SourceRange>,
+            keyword: ParseKeyword,
+        }
+        impl Node for $name {
+            fn kind(&self) -> Kind<'_> {
+                Kind::Keyword(self)
+            }
+        }
+        impl Keyword for $name {
+            fn new(range: SourceRange, keyword: ParseKeyword) -> Self {
+                Self {
+                    range: Some(range),
+                    keyword,
+                }
+            }
+            fn keyword(&self) -> ParseKeyword {
+                self.keyword
+            }
+            fn allowed_keywords() -> &'static [ParseKeyword] {
+                &[$(ParseKeyword::$allowed),*]
+            }
+            fn as_leaf(&self) -> &dyn Leaf {
+                self
+            }
+        }
+        impl TryParse for $name {
+            fn try_parse(pop: &mut Populator) -> ParseResult<Self> {
+                pop.parse_keyword()
+            }
+        }
+    }
+}
+
+/// Define a node that implements the token trait.
+macro_rules! define_token_node {
+    ( $name:ident, $($allowed:ident),* $(,)? ) => {
+        #[derive(Default, Debug, Leaf!)]
+        pub struct $name {
+            range: Option<SourceRange>,
+            parse_token_type: ParseTokenType,
+        }
+        impl Node for $name {
+            fn kind(&self) -> Kind<'_> {
+                Kind::Token(self)
+            }
+        }
+        impl Token for $name {
+            fn new(range: SourceRange, token_type: ParseTokenType) -> Self {
+                Self {
+                    range: Some(range),
+                    parse_token_type: token_type,
+                }
+            }
+            fn token_type(&self) -> ParseTokenType {
+                self.parse_token_type
+            }
+            fn allowed_tokens() -> &'static [ParseTokenType] {
+                Self::ALLOWED_TOKENS
+            }
+            fn as_leaf(&self) -> &dyn Leaf {
+                self
+            }
+        }
+        impl CheckParse for $name {
+            fn can_be_parsed(pop: &mut Populator<'_>) -> bool {
+                let typ = pop.peek_type(0);
+                Self::ALLOWED_TOKENS.contains(&typ)
+            }
+        }
+        impl Parse for $name {
+            fn parse(pop: &mut Populator<'_>) -> Self {
+                pop.parse_token()
+            }
+        }
+        impl $name {
+            const ALLOWED_TOKENS: &'static [ParseTokenType] = &[$(ParseTokenType::$allowed),*];
+        }
+    }
+}
+
+macro_rules! define_list_nodes {
+    (
+        // due to macro hygiene, the caller needs to provide their own
+        // variable name for the populator if they intend to access it
+        // just specify it up here, and all the trait impls are free to use it if they need
+        let $pop:ident;
+
+        $(
+            $contents:ty => $name:ident {
+                chomp_newlines: $newlines:expr,
+                chomp_semis: $semis:expr,
+                $(stop_unwind: $unwind:expr,)?
```

Comment:
The name of this is `let $pop_:ident;` in an earlier commit, which
is confusing to read. Can you fix that, i.e. use the better name from
the start?